### ISSUE TITLE
Post setup Python step fails (pipenv cache)

### ISSUE THREAD

**Issue by mrlubos** [actions/setup-python#436]
**Description:**

Hey! So this used to work just a few days ago. We’ve been having some issues with cache in the last few days (see [here for details](https://stackoverflow.com/questions/72598852/getcacheentry-failed-cache-service-responded-with-503)), this is a brand new issue which I think might or might not be related to the previous problem. Simply put, the action fails in the post step with the following message.

```sh
Post job cleanup.
[2](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:20:2)
Error: Cache folder path is retrieved for pipenv but doesn't exist on disk: /home/runner/.local/share/virtualenvs
```

Here’s the actual step:

```sh
Run actions/setup-python@v3.1.2
[2](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:5:2)
  with:
[3](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:5:3)
    cache: pipenv
[4](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:5:4)
    python-version: 3.9.13
[5](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:5:5)
    token: ***
[6](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:5:7)
Successfully setup CPython (3.9.13)
[7](https://github.com/<org>/<repo>/runs/6889538635?check_suite_focus=true#step:5:8)
pipenv cache is not found
```

**Action version:**
Specify the action version

3.1.2

**Platform:**
- [x] Ubuntu
- [ ] macOS
- [ ] Windows

**Runner type:**
- [x] Hosted
- [ ] Self-hosted

**Tools version:**

3.9.13

**Repro steps:**  
A description with steps to reproduce the issue. If your have a public example or repo to share, please provide the link.

**Expected behavior:**
A description of what you expected to happen.

No issue occurs, the same way it used to work a week ago.

**Actual behavior:**
A description of what is actually happening.

This error message appears on every run.

**Comment by dmitry-shibanov** [comment]
Hello @mrlubos. Thank you for your report. I've tried to reproduce the issue but everything works as expected. Could you please provide public repository to reproduce the issue ?

**Comment by mrlubos** [comment]
@dmitry-shibanov Hi Dmitry, thank you for your reply! I might look into trying to reproduce this in a public repository if the issue persists, wanted to provide at least some pointers in the meantime in case anyone else runs into the same problem. Further behaviour I observed today is that even if the post setup step completes, the next run will show this message:

```sh
Run actions/setup-python@v3.1.2
  with:
    cache: pipenv
    python-version: [3](https://github.com/<org>/<repo>/runs/6896643828?check_suite_focus=true#step:5:3).9.13
    token: ***
Successfully setup CPython (3.9.13)
pipenv cache is not found
```

What I think might get us closer to the resolution is the fact I realised we've started experiencing these issues after upgrading from v2 to v3. That being said, the post step cache looks the same for both versions:

v2
```sh
Post job cleanup.
Cache hit occurred on the primary key setup-python-Linux-python-3.9.13-pipenv-cbf9160[2](https://github.com/<org>/<repo>/runs/6766531090?check_suite_focus=true#step:18:2)ee5daeb2e69e4e65d32ca4c617d796670afbc226f170e9e4035d058e, not saving cache.
```

v3
```sh
Post job cleanup.
/usr/bin/tar --posix --use-compress-program zstd -T0 -cf cache.tzst -P -C /home/runner/work/<repo>/<repo> --files-from manifest.txt
Cache Size: ~[2](https://github.com/<org>/<repo>/runs/6896643828?check_suite_focus=true#step:18:2)02 MB (212215749 B)
Cache saved successfully
Cache saved with the key: setup-python-Linux-python-[3](https://github.com/<org>/<repo>/runs/6896643828?check_suite_focus=true#step:18:3).9.13-pipenv-0c70[4](https://github.com/<org>/<repo>/runs/6896643828?check_suite_focus=true#step:18:4)b[5](https://github.com/<org>/<repo>/runs/6896643828?check_suite_focus=true#step:18:5)9407d6fda6f2722fa4daa27bc67f9d74da6402ee4c95f412a71936434
```

I would therefore expect the cache to be picked up by v3. I am shooting blanks here, would adding `fetch-depth: 0` to the checkout action help by any chance?

**Comment by dmitry-shibanov** [comment]
Thank you for your reply @mrlubos. I do not think that it's related to fetch depth. I'll try to reproduce the issue from my side one more time. If you have any updates feel free to post them here.

**Comment by mrlubos** [comment]
Hi @dmitry-shibanov no updates beyond that this is still an issue. I tried removing the `cache` key which obviously removes the error but after adding it back, the error came back too. So I am stuck it seems between not having caching or receiving this error. I don't have this confirmed yet, but I suspect this error happens only on the `master` branch and not in other branches. Also this is interesting to me, after I added back the `cache` key, the next run produced this:

```sh
Run actions/setup-python@v3.1.2
  with:
    cache: pipenv
    python-version: 3.9.13
    token: ***
Successfully setup CPython (3.9.13)
pipenv cache is not found
```

And then in post step

```sh
Post job cleanup.
Error: Cache folder path is retrieved for pipenv but doesn't exist on disk: /home/runner/.local/share/virtualenvs
```

**Comment by mrlubos** [comment]
@dmitry-shibanov So I am fairly confident this is happening only on the `master` branch. All runs on other branches complete without issues, it's happening every time on `master` though

**Comment by coldino** [comment]
I was also seeing this when attempting to update to v4 (from v1!).

After some head scratching, I found it was because of some remnants of my old manual caching system:

```yaml
env:
  PIPENV_NOSPIN: 1
  PIPENV_CACHE_DIR: /tmp/.pip-cache
  WORKON_HOME: /tmp/pipenv-venvs
```

Removing both cache location vars enables the "new" caching system to work as expected.

**Comment by mrlubos** [comment]
Yeah @coldino it does sound like something similar is happening to me, except I don't have any variables set explicitly so it's not obvious whether this is something I can resolve on my end

**Comment by lsmith77** [comment]
I am also seeing this issue. I upgraded to `@v4` a while ago but I am only seeing this issue now that I changed a `pipenv` dependency

<img width="718" alt="image" src="https://user-images.githubusercontent.com/300279/201078428-590739d0-c86c-46d3-b9fb-973aec734d03.png">

This here is the relevant portion of my github actions setup

```
name: run tests

on: [push]

jobs:
  build:
    runs-on: ubuntu-latest
    env:
      PIPENV_VENV_IN_PROJECT: enabled

    steps:
      - uses: actions/checkout@v3
        with:
          fetch-depth: 0

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v4
        with:
          python-version: "3.9"
          cache: "pipenv"

      - name: Install pipenv
        run: |
          python -m pip install --upgrade pipenv wheel

      - name: Install dependencies
        run: |
          pipenv install --deploy --dev --verbose
```

**Comment by lsmith77** [comment]
removing `PIPENV_VENV_IN_PROJECT: enabled` seems to have fixed things for me.

**Comment by MemphisMeng** [comment]
I figured that this issue persisted even when the name is `main` instead of `master`.

**Comment by matteius** [comment]
FWIW -- about a month ago pipenv changed to be more strict on if `PIPENV_VENV_IN_PROJECT` is set to one of these truthy values:  https://github.com/pypa/pipenv/blob/main/pipenv/utils/constants.py#L5

Interestingly though, we ran into this issue today as well at work and found we also had `PIPENV_VENV_IN_PROJECT = enabled`, but after changing it to `PIPENV_VENV_IN_PROJECT = 1` the issue is still obtained.  I help maintain pipenv in my spare time, and this is odd behavior as we generally expect that folks can use `PIPENV_VENV_IN_PROJECT = 1` in their github CIs.  I'll have to think more about what could be causing this issue.

**Comment by matteius** [comment]
It would appear the pipenv cache implementation assumes it will be at this location, making `PIPENV_VENV_IN_PROJECT` probably unworkable at present versions of setup-python actions:  https://github.com/actions/setup-python/blob/main/src/cache-distributions/pipenv-cache.ts#L25

**Comment by rburema** [comment]
At the risk of sounding _very_ obvious, but maybe this'll help someone in any case:

We ran into this issue recently. (For example; [see here](https://github.com/Ultimaker/Cura/actions/runs/4197625247/jobs/7280187388).)

What ['fixed'](https://github.com/Ultimaker/Cura/pull/14581) it for us was the realization that whenever this issue occurred, the workflow as a whole needed to skip all the steps that actually required python and/or pip -- and then to disable the setup-python step as well whenever that was the case.

(That might also be going on with the reproduction of others' but in a more subtle way: All of the actions taken that I can see here are for packages that are likely to be included and at latest version anyway, so the cache might never get written to. Again, sorry if that was already obvious.)

In any case, however inelegant it may be, setting up something, not using it, and then cleaning it up again should probably (also) not result in an error. (Just putting that out there, since that is a slightly different case than what OP is talking about, which may or may not require a different approach to solve.)

**Comment by tiangolo** [comment]
I'm seeing the same error in FastAPI, with `pip`, not pipenv, here's an example of one of the runs: https://github.com/tiangolo/fastapi/actions/runs/4777885541/jobs/8541931450

In the initial stage of the run I get:

```
Run actions/setup-python@v4
Installed versions
/opt/hostedtoolcache/Python/3.9.16/x64/bin/pip cache dir
/home/runner/.cache/pip
pip cache is not found
```

And then in the Post Setup I get:

```
Post job cleanup.
Error: Cache folder path is retrieved for pip but doesn't exist on disk: /home/runner/.cache/pip
```

**Comment by BR0kEN-** [comment]
I have two repos with the exact same job each.

```yaml
      - uses: actions/setup-python@v4
        with:
          python-version: "3.8"
          cache: pip
```

In fact, only Python is needed while no PIP packages are installed so there is nothing to cache.

For some reason, one repo outputs:

```
Post job cleanup.
Cache hit occurred on the primary key setup-python-Linux-20.04-Ubuntu-python-3.8.17-pip-d7503bb43f0a5d313eaed44aef3115b7278b36b13cd6e4365bddce2086d3e43a, not saving cache.
```

While another one (on every run):

```
Post job cleanup.
Error: Cache folder path is retrieved for pip but doesn't exist on disk: /home/runner/.cache/pip
```

**Comment by simonw** [comment]
I'm seeing this error too.

**Comment by Amet13** [comment]
Same for me.

```
      - uses: actions/setup-python@v4
        with:
          python-version: 3.11
          cache: 'pipenv'
```

```
Error: Cache folder path is retrieved for pipenv but doesn't exist on disk: /home/runner/.local/share/virtualenvs
```

**Comment by MSK61** [comment]
I have the same error with a pipenv cache setting.

**Comment by doc-sheet** [comment]
@smougenot, I guess in your case it's better to call `mkdir -p /home/runner/.local/share/virtualenvs` directly.
looks like pipenv's cache location is different from pip

**Comment by woodruffw** [comment]
I've experienced another variant of this: if you add `actions/setup-python` to a workflow job and enable caching, but you job *doesn't actually use* Python or fill the cached directory in any ways, then the post-step cleanup will fail with this message.

Example failing run:

https://github.com/C2SP/x509-limbo/actions/runs/9739572591/job/26897420759

Exact reproduction steps:

1. Have no cache to restore from
2. Use `actions/setup-python` with `cache: pip`
3. Pre-step succeeds (but does nothing, as there's no cache to restore from)
4. Post-step fails (the cache wasn't modified during the job at all, so it isn't present)

The failure in (4) seems unambiguously incorrect to me: a caching layer should never fail on an empty cache if no cache was loaded to begin with.

**Comment by aparnajyothi-y** [comment]
Hello Everyone, Thank you once again for creating this issue and  The error related to pipenv or poetry package managers cannot be changed to a warning due to the necessity of having the Pipfile.lock or poetry.lock file for caching in the actions/Setup-python.

To resolve this issue, please consider one of the following options:

Add the Pipfile.lock / poetry.lock file: Ensure that the Pipfile.lock file or poetry.lock is present in your repository. You can generate this file by running pipenv lock.
Install dependencies: Make sure all required dependencies are installed and committed.
Do not use cache: pip: If there are no dependencies to cache, remove the cache: pip configuration from your GitHub Actions workflow.
Additionally, please note that changing this behaviour would require multiple code changes, which may be implemented as a feature request.

**Comment by aparnajyothi-y** [comment]
Hello @mrlubos, Please let us know if you have any further concerns on the above information.

**Comment by woodruffw** [comment]
> Do not use cache: pip: If there are no dependencies to cache, remove the cache: pip configuration from your GitHub Actions workflow.

I want to push back slightly on this: caching should always be referentially transparent, meaning that the cache step *itself* should never fail, even if the caching operation fails or can't be completed (e.g. if there's nothing to cache).

With that being said, I can understand that this would require multiple changes to fix. As a stop-gap, my suggestion would be to improve the error messaging here: if the caching step fails because there's nothing to cache, the error message could say that explicitly and suggest that the maintainer remove the step entirely, rather than returning the current (confusing) message.

**Comment by insanitybit** [comment]
It would be nice to fail gracefully here. Sometimes an action will install deps, sometimes it won't - right now this bug means you have to completely forgo caching via `pip`.

**Comment by aparnajyothi-y** [comment]
Hello Everyone, The [PR #968](https://github.com/actions/setup-python/pull/968) has been merged to update the error message when no dependencies are found to cache, improving its clarity.

Please validate the changes by using the actions/setup-python@main branch and confirm to proceed to close this issue.

**Comment by woodruffw** [comment]
That error message looks better, thank you! However, I'd like to reiterate that I think this shouldn't be an error at all, per caching semantics. As such I think this should remain open until the post-step actually ceases to fail.

**Comment by aparnajyothi-y** [comment]
Hello @mrlubos, @woodruffw,

Please let us know if the above information has clarified your concerns regarding the throwing an error message at point of the post-setup phase. We're happy to provide any further assistance if needed!

**Comment by woodruffw** [comment]
> Throwing an error message at this point in the post-setup phase is crucial for handling intermittent failures during package installations. Dependencies can sometimes fail to install due to issues such as network instability, timeouts, or conflicts. By raising an error at this stage, we can explicitly identify and address these failures, rather than letting the workflow proceed with an incomplete or inconsistent environment.

These are all good reasons why the post-setup step should fail. But these are different reasons than *cache restoration*, which itself should never fail, since a cache is meant to be a transparent layer. 

In other words: I think it makes sense for the post-setup to fail if a package installation fails, but the presence of an empty cache state is not *itself* evidence of a failure.

**Comment by aparnajyothi-y** [comment]
Hello @woodruffw, Thank you for your thoughtful response. We completely understand your perspective and agree that cache restoration should ideally be a transparent process. However, we believe that throwing an error message during the post-setup phase is critical for effectively handling intermittent failures, particularly during package installations.

While the cache itself might not indicate a failure, the absence of expected package data in the cache can still disrupt the build process. By throwing an error in such cases, we can better identify and address issues arising from these intermittent failures, ensuring that the setup process remains both reliable and predictable.

We appreciate your input and hope this clarifies our approach. 

Please let us know if you have any further concerns or clarifications.

**Comment by woodruffw** [comment]
> While the cache itself might not indicate a failure, the absence of expected package data in the cache can still disrupt the build process. By throwing an error in such cases, we can better identify and address issues arising from these intermittent failures, ensuring that the setup process remains both reliable and predictable.

I don't follow this reasoning, sorry -- `pip` and other Python packaging tooling are resilient to partial caches, and shouldn't fail just because something wasn't cached (because this again would violate the semantics of caching). The build process should *never* be disrupted by a cache miss, since `pip` et al. will hit the index if a miss occurs. 

Do you have a specific example of a Python packaging tool or process that fails because of a partial or empty cache?

**Comment by JamesGuthrie** [comment]
I agree with @woodruffw here. I don't understand why "you didn't use the cache" is a valid reason to throw an error in the post setup step.

**Comment by v-aparnajyothi-y** [comment]
Hello Everyone, Thank you for the thoughtful feedback on this! We agree that cache misses should not be treated as fatal errors, especially since caching is intended to be a performance optimization rather than a hard requirement. Based on both product direction and community input including this discussion, we are planning to downgrade post-setup cache failures from errors to warnings.

This change ensures that workflows continue even when a cache is missing or incomplete, aligning with the expectation that caching should remain transparent and non-blocking. We’ll continue to monitor for any unintended side effects and aim to strike the right balance between resilience and visibility into real issues.

If you have any concerns or feedback following this change, please don’t hesitate to reach out. we’d really appreciate it!

**Comment by v-aparnajyothi-y** [comment]
Hello Everyone, This issue has been addressed in [PR #1182](https://github.com/actions/setup-python/pull/1182). To utilize this feature, you can use the actions/setup-python@main branch in your workflows for now, and the changes will be included in the upcoming releases.

We’ll be closing this issue. Please feel free to reach out if you have any further concerns or need additional assistance. Thanks again for your suggestion and feedback!
